### PR TITLE
Pso fix

### PR DESCRIPTION
**Pull request by hellkite500**
Quick fix to workaround multiprocessing/pickle interactions with the plugin manager for PSO optimization.  Also relevant to #135.

### CODE REVIEW

File: python/ngen_cal/src/ngen/cal/_plugin_system.py [hunk 1]
```diff
@@ -27,3 +27,17 @@ def setup_plugin_manager(plugins: list[Callable | ModuleType]) -> PluginManager:
             assert_never(plugin)
 
     return pm
+
+def setup_scoped_plugin_manager(spec: type, plugins: list[Callable | ModuleType]) -> PluginManager:
```
Comment (aaraney) [inline python/ngen_cal/src/ngen/cal/_plugin_system.py:31]:
```suggestion
def setup_namespaced_plugin_manager(spec: type, plugins: list[Callable]) -> PluginManager:
```

I think the type hints here can be cleaned up slightly just emphasize the invariants at play.

File: python/ngen_cal/src/ngen/cal/_plugin_system.py [hunk 2]
```diff
@@ -27,3 +27,17 @@ def setup_plugin_manager(plugins: list[Callable | ModuleType]) -> PluginManager:
...
+    pm = pluggy.PluginManager(PROJECT_SLUG)
+    pm.add_hookspecs(spec)
+
+    for plugin in plugins:
+        if isinstance(plugin, Callable):
+            pm.register(plugin())
+        elif isinstance(plugin, ModuleType):
```
Comment (aaraney) [inline python/ngen_cal/src/ngen/cal/_plugin_system.py:38]: If a module based plugin is registered, we will run into the same pickling a module issue in a multiprocessing context. I think we should just `assert not isinstance(plugin, ModuleType), "function plugins"` and catch that at the call site and re-raise with extra context.

File: python/ngen_cal/src/ngen/cal/model.py [hunk 3]
```diff
@@ -233,7 +234,7 @@ class Config(BaseModel.Config):
     def __init__(self, *args, **kwargs):
         super().__init__(**kwargs)
         model_plugins = cast(List[Union[Callable, ModuleType]], self.plugins)
-        self._plugin_manager = setup_plugin_manager(model_plugins)
+        self._plugin_manager = setup_scoped_plugin_manager(ModelHooks, model_plugins)
```
Comment (aaraney) [inline python/ngen_cal/src/ngen/cal/model.py:237]:
```suggestion
        try:
            self._plugin_manager = setup_scoped_plugin_manager(ModelHooks, model_plugins)
        except AssertionError as e:
            raise RuntimeError("At this time, model plugins do not support function based plugins.")
```

File: python/ngen_cal/src/ngen/cal/_plugin_system.py (hunk 1 above)
Comment (hellkite500) [reply python/ngen_cal/src/ngen/cal/_plugin_system.py:30]: changed in a4710a941b0869d1bf1c2d2bfd920b93b09646fc

File: python/ngen_cal/src/ngen/cal/_plugin_system.py (hunk 2 above)
Comment (hellkite500) [reply python/ngen_cal/src/ngen/cal/_plugin_system.py]: refactored in a4710a941b0869d1bf1c2d2bfd920b93b09646fc

**Review by aaraney — APPROVED** [review]
Thanks, @hellkite500!

**Review by aaraney — CHANGES_REQUESTED** [review]
Looks like one of the tests if correctly failing now.

https://github.com/NOAA-OWP/ngen-cal/actions/runs/10096240522/job/27918258512?pr=137#step:7:91

**Comment by aaraney** [comment]
Looks like these test function plugins need to be removed now (https://github.com/NOAA-OWP/ngen-cal/blob/61c9672559832d5bb2edf7471f8d1c1cf959f9c6/python/ngen_cal/tests/test_plugin_system.py#L53-L59).

**Comment by hellkite500** [comment]
The problem was that I modifed `setup_plugin_manager` instead of `setup_scoped_plugin_manager` haha, I just corrected that.

**Comment by aaraney** [comment]
> The problem was that I modifed `setup_plugin_manager` instead of `setup_scoped_plugin_manager` haha, I just corrected that.

Nice!

**Comment by aaraney** [comment]
In doing some local testing, we should be able to get around some of this if we use [`multiprocess`](https://github.com/uqfoundation/multiprocess?tab=readme-ov-file#multiprocess). It uses [`dill`](https://github.com/uqfoundation/dill) an alternative to pickle, to serialize and deserialize objects.

**Pull request merged**